core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+                'version': item.exploration_version,
+                'session_id': item.session_id
+            })
+
+        elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+            value = {
+                'event_type': feconf.EVENT_TYPE_STATE_HIT,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'session_id': item.session_id,
+                'created_on': utils.get_time_in_millisecs(item.created_on)
+            }
+            yield (item.exploration_id, value)
+
+        else:
+            value = {
+                'event_type': GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS,
+                'version': item.exploration_version,
+                'state_name': item.state_name,
+                'submitted_answer_list': item.submitted_answer_list
+            }
+            yield (item.exploration_id, value)
+
+    @staticmethod
+    def reduce(exp_id, stringified_values):
+        num_starts = 0
+        num_completions = 0
+        num_actual_starts = 0
+
+        exploration = exp_services.get_exploration_by_id(exp_id)
+        latest_exp_version = exploration.version
+
+        # The list of exploration start events in stringified_values.
+        start_exploration_event_dicts = []
+        # The list of exploration complete events in stringified_values.
+        complete_exploration_event_dicts = []
+        # The list of state hit events in stringified_values.
+        state_hit_event_dicts = []
+        # The list of state answers instances in stringified_values.
+        state_answer_event_dicts = []
+        # The list of session_ids of learners completing an exploration.
+        completed_session_ids = []
+        for value_str in stringified_values:
+            value = ast.literal_eval(value_str)
+
+            if value['event_type'] == feconf.EVENT_TYPE_START_EXPLORATION:
+                start_exploration_event_dicts.append(value)
+            elif value['event_type'] == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
+                complete_exploration_event_dicts.append(value)
+                completed_session_ids.append(value['session_id'])
+            elif value['event_type'] == feconf.EVENT_TYPE_STATE_HIT:
+                state_hit_event_dicts.append(value)
+            elif value['event_type'] == (
+                    GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS):
+                state_answer_event_dicts.append(value)
+
+        # This dict will store the cumulative values of state stats from version
+        # 1 till the latest version of an exploration.
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of timestamp.
+        state_hit_event_dicts = sorted(state_hit_event_dicts, key=lambda e: e[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            start_exploration_events_for_this_version = [
```

Comment:
This is a nested for loop. Don't do this. Group by version at the outset. Otherwise your code is going to time out in production (we have millions of events).

Ditto below.